File: problems/Swift/700.二叉搜索树中的搜索.swift
/**
 * 700. 二叉搜索树中的搜索 [Easy]
 * Time Complexity: O(n), Space Complexity: O(n)
 * https://leetcode-cn.com/problems/search-in-a-binary-search-tree/
 */

/**
 * Definition for a binary tree node.
 */
public class TreeNode {
  public var val: Int
  public var left: TreeNode?
  public var right: TreeNode?
  public init() { val = 0; left = nil; right = nil }
  public init(_ val: Int) { self.val = val; left = nil; right = nil }
  public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
    self.val = val
    self.left = left
    self.right = right
  }
}

class Solution {
  func searchBST(_ root: TreeNode?, _ val: Int) -> TreeNode? {
    guard let root = root else { return nil }

    if root.val == val {
      return root
    }

    return searchBST(val < root.val ? root.left : root.right, val)
  }
}

File: problems/rust-solutions/src/700.二叉搜索树中的搜索.rs
// 700.二叉搜索树中的搜索 [Easy]
// Time Complexity: O(n), Space Complexity: O(n)
// https://leetcode-cn.com/problems/search-in-a-binary-search-tree/submissions/

// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
  pub val: i32,
  pub left: Option<Rc<RefCell<TreeNode>>>,
  pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
  #[inline]
  pub fn new(val: i32) -> Self {
    TreeNode {
      val,
      left: None,
      right: None,
    }
  }
}
use std::cell::RefCell;
use std::rc::Rc;

impl Solution {
  pub fn search_bst(
    root: Option<Rc<RefCell<TreeNode>>>,
    val: i32,
  ) -> Option<Rc<RefCell<TreeNode>>> {
    if root.is_none() {
      return None;
    }

    if root.as_ref().unwrap().borrow().val == val {
      return root;
    }
    let node = Solution::search_bst(root.as_ref().unwrap().borrow().left.clone(), val);
    if node.is_some() {
      return node;
    } else {
      return Solution::search_bst(root.as_ref().unwrap().borrow().right.clone(), val);
    }
  }
}
